File: src/dotnet/AdventOfCode2021/AdventOfCode2021.Solutions/Day7.cs
namespace AdventOfCode2021.Solutions;

public partial class Day7 : IDay
{
    private record struct Crab(int Position, int Count);

    private record struct Travel(Crab Crab, int Distance);

    private record struct Result(int POsition, long TotalFuel);

    public int DayNumber => 7;

    public long CalculatePartOne()
    {
        return CalculateMinimumFuel(x => x);
    }

    public long CalculatePartTwo()
    {
        return CalculateMinimumFuel(x => (x + 1) * x / 2);
    }

    private static long CalculateMinimumFuel(Func<int, int> fuelCalculator)
    {
        var input = ParseInput();

        var minPosition = input.Min();
        var maxPosition = input.Max();

        var crabs = input.GroupBy(x => x).Select(x => new Crab(x.Key, x.Count())).ToArray();

        var result = Enumerable.Range(minPosition, maxPosition - minPosition + 1)
            .Select(pos =>
            {
                var fuel = crabs.Select(cr => new Travel(cr, Math.Abs(cr.Position - pos)))
                            .Aggregate(0L, (fuel, crabs) => fuel + fuelCalculator(crabs.Distance) * crabs.Crab.Count);
                return new Result(pos, fuel);
            }).OrderBy(r => r.TotalFuel).First();

        return result.TotalFuel;
    }

    private static int[] ParseInput()
    {
        return Input.Split(",").Select(int.Parse).ToArray();
    }
}

File: src/kotlin/app/src/main/kotlin/ru/renev/adventofcode2021/Day7.kt
package ru.renev.adventofcode2021

import kotlin.math.abs

class Day7 {
    data class Crab(val position: Int, val count: Int)

    data class Travel(val crab: Crab, val distance: Int)

    data class Result(val position: Int, val totalFuel: Long)

    fun part1(): Long {
        println("Day 7 part 1")

        return calculateMinimumFuel { it }
    }

    fun part2(): Long {
        println("Day 7 part 2")

        return calculateMinimumFuel { (it + 1) * it / 2 }
    }

    private fun calculateMinimumFuel(fuelCalculator: (distance: Int) -> Int): Long {
        val input = parseInput()

        val minPosition = input.minOf { it }
        val maxPosition = input.maxOf { it }

        val crabs = input.groupBy { it }
            .map { Crab(it.key, it.value.count()) }

        val result = (minPosition..maxPosition).map { position ->
            crabs.map {
                Travel(it, abs(it.position - position))
            }
                .fold(0L) { sum, pos ->
                    sum + fuelCalculator(pos.distance) * pos.crab.count
                }
                .let { Result(position, it) }
        }.maxByOrNull { it.totalFuel }!!

        return result.totalFuel
    }

    private fun parseInput(): List<Int> {
        return Day7Input.values.split(",").map { it.toInt() }
    }
}
